tests/interfaces/test_batch_torch.py

```diff
@@ -0,0 +1,875 @@
+
+        res = torch.autograd.functional.hessian(cost_fn, params)
+        expected = torch.tensor(
+            [
+                [-np.cos(2 * x) * np.cos(2 * y), np.sin(2 * x) * np.sin(2 * y)],
+                [np.sin(2 * x) * np.sin(2 * y), -2 * np.cos(x) ** 2 * np.cos(2 * y)],
+            ]
+        )
+        assert torch.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_hessian_vector_valued(self, tol):
+        """Test hessian calculation of a vector valued QNode"""
+        dev = qml.device("default.qubit", wires=1)
+
+        def circuit(x):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(x[0], wires=0)
+                qml.RX(x[1], wires=0)
+                qml.probs(wires=0)
+
+            return torch.stack(execute([tape], dev, gradient_fn=param_shift, interface="torch"))
+
+        x = torch.tensor([1.0, 2.0], requires_grad=True)
+        res = circuit(x)
+
+        a, b = x.detach().numpy()
+
+        expected_res = [
+            0.5 + 0.5 * np.cos(a) * np.cos(b),
+            0.5 - 0.5 * np.cos(a) * np.cos(b),
+        ]
+        assert np.allclose(res.detach(), expected_res, atol=tol, rtol=0)
+
+        jac_fn = lambda x: torch.autograd.functional.jacobian(circuit, x, create_graph=True)
+
+        g = jac_fn(x)
+
+        hess = torch.autograd.functional.jacobian(jac_fn, x)
+
+        expected_g = [
+            [-0.5 * np.sin(a) * np.cos(b), -0.5 * np.cos(a) * np.sin(b)],
+            [0.5 * np.sin(a) * np.cos(b), 0.5 * np.cos(a) * np.sin(b)],
+        ]
+        assert np.allclose(g.detach(), expected_g, atol=tol, rtol=0)
+
+        expected_hess = [
+            [
+                [-0.5 * np.cos(a) * np.cos(b), 0.5 * np.sin(a) * np.sin(b)],
+                [0.5 * np.sin(a) * np.sin(b), -0.5 * np.cos(a) * np.cos(b)],
+            ],
+            [
+                [0.5 * np.cos(a) * np.cos(b), -0.5 * np.sin(a) * np.sin(b)],
+                [-0.5 * np.sin(a) * np.sin(b), 0.5 * np.cos(a) * np.cos(b)],
+            ],
+        ]
+        assert np.allclose(hess.detach(), expected_hess, atol=tol, rtol=0)
+
+    def test_adjoint_hessian(self, tol):
+        """Since the adjoint hessian is not a differentiable transform,
+        higher-order derivatives are not supported."""
+        dev = qml.device("default.qubit", wires=2)
+        params = torch.tensor([0.543, -0.654], requires_grad=True, dtype=torch.float64)
+
+        def cost_fn(x):
+            with qml.tape.JacobianTape() as tape:
+                qml.RX(x[0], wires=[0])
+                qml.RY(x[1], wires=[1])
+                qml.CNOT(wires=[0, 1])
+                qml.expval(qml.PauliZ(0))
+
+            return execute(
+                [tape],
+                dev,
+                gradient_fn="device",
+                gradient_kwargs={"method": "adjoint_jacobian", "use_device_state": True},
+                interface="torch",
+            )[0]
+
+        res = torch.autograd.functional.hessian(cost_fn, params)

```

Comment:
If higher derivatives aren't supported I would expect this to through an error rather than just returning 0s (maybe I'm missing something though?)

Reply:
Let me see if I can fix this 🤔

PyTorch provides a [`mark_non_differentiable`](https://pytorch.org/docs/stable/generated/torch.autograd.function._ContextMethodMixin.mark_non_differentiable.html#torch.autograd.function._ContextMethodMixin.mark_non_differentiable) decorator that I tried using in the past, to no avail. Even when it was present, it would still 'let' you differentiate further (but would just return zeros).

The same thing occurs in Autograd (it returns zeros), but they nicely also raise a warning ('Output seems independent of input?') which is nice 😆

Reply:
Yeah I think a warning at least would be great!

Reply:
@glassnotes I tried `mark_non_differentiable`, but it still doesn't seem to have any effect :( PyTorch still acts like everything is working, and the output is just 0s.

Unfortunately, this is not something I think we can fix in PennyLane, since it is an internal implementation detail of PyTorch!

Reply:
Oh darn! Thanks for looking into it :+1: